// markmap-lib/dist/plugins/prism.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.transform = transform;
exports.name = void 0;

var _prismjs = _interopRequireDefault(require("prismjs"));

var _components = _interopRequireDefault(require("prismjs/components/"));

const name = 'prism';
exports.name = name;

function transform(transformHooks) {
  transformHooks.parser.tap((md, features) => {
    md.set({
      highlight: (str, lang) => {
        features[name] = true;
        let grammar = _prismjs.default.languages[lang];

        if (!grammar) {
          (0, _components.default)([lang]);
          grammar = _prismjs.default.languages[lang];
        }

        if (grammar) {
          return _prismjs.default.highlight(str, grammar, lang);
        }

        return '';
      }
    });
  });
  return {
    styles: [{
      type: 'stylesheet',
      data: {
        href: 'https://cdn.jsdelivr.net/npm/prismjs@1/themes/prism.css'
      }
    }]
  };
}
