// markmap-lib/dist/template.js
"use strict";

exports.__esModule = true;
exports.fillTemplate = fillTemplate;

var _util = require("./util");

const template = "<!DOCTYPE html>\n<html>\n<head>\n<meta charset=\"UTF-8\">\n<meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">\n<meta http-equiv=\"X-UA-Compatible\" content=\"ie=edge\">\n<title>Markmap</title>\n<style>\n* {\n  margin: 0;\n  padding: 0;\n}\n#mindmap {\n  display: block;\n  width: 100vw;\n  height: 100vh;\n}\n</style>\n<!--CSS-->\n</head>\n<body>\n<svg id=\"mindmap\"></svg>\n<!--JS-->\n</body>\n</html>\n";
const version = "0.9.3";
const baseJs = ['https://cdn.jsdelivr.net/npm/d3@5', `https://cdn.jsdelivr.net/npm/markmap-lib@${version}/dist/browser/view.min.js`].map(src => ({
  type: 'script',
  data: {
    src
  }
}));

function fillTemplate(data, opts) {
  const {
    scripts,
    styles
  } = opts;
  const cssList = [...(styles ? (0, _util.persistCSS)(styles) : [])];
  const context = {
    getMarkmap: () => window.markmap,
    data
  };
  const jsList = [...(0, _util.persistJS)(baseJs), ...(scripts ? (0, _util.persistJS)(scripts, context) : []), ...(0, _util.persistJS)([{
    type: 'iife',
    data: {
      fn: (getMarkmap, data) => {
        const {
          Markmap
        } = getMarkmap();
        window.mm = Markmap.create('svg#mindmap', null, data);
      },
      getParams: ({
        getMarkmap,
        data
      }) => {
        return [getMarkmap, data];
      }
    }
  }], context)];
  const html = template.replace('<!--CSS-->', () => cssList.join('')).replace('<!--JS-->', () => jsList.join(''));
  return html;
}
